core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -259,6 +258,14 @@ public static State fromString(String state) {
                       Setting.Property.Dynamic,
                       Setting.Property.IndexScope);
 
+    /**
+     * an internal index format description, allowing us to find out if this index is upgraded or needs upgrading
+     */
+    private static final String INDEX_INTERNAL_FORMAT = "index.internal.format";
+    public static final Setting<Integer> INDEX_INTERNAL_FORMAT_SETTING =
+            Setting.intSetting(INDEX_INTERNAL_FORMAT, 0, Setting.Property.IndexScope);
+    static final int INTERNAL_INDEX_FORMAT_CURRENT = 6;
```

Comment:
maybe we make this a bit more generic and call it `index.format` and `INTERNAL_INDEX_FORMAT_CURRENT` should not be here. it's up to every user of this setting to define the current?